### PR TITLE
Debug console enhancements

### PR DESCRIPTION
**Pull request by islemaster**
Extract debug console history into a class with a few tests, and enhance the console in a couple of ways:
- Limit console command history to 64 commands, so we don't have an ever-increasing array.
- Make clicking on the (noninteractive) console output select the console input, if no text was selected.

Implements improvements originally written by @ty-po in #3230 - since I've recently refactored the debug console pretty far away from his original work, I'm going ahead and merging his work myself.

### CODE REVIEW

**Comment by islemaster** [comment]
FYI @davidsbailey 

Brian, I'm reintroducing mixins.scss since I have another use for it.  I'm renaming the mixin so it doesn't override `user-select` in case that's part of the original issue with sprockets that you fixed.  I don't know how this has anything to do with sprockets now... the affected scss files are build as part of the apps build process before Rails touches them, and I've verified that the build output is valid.  Do you see any problems with this?

@ty-po please review as well, since you did the original implementation of these features!

File: apps/test/CommandHistoryTest.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+describe("CommandHistory", function () {
+  var CommandHistory = require('@cdo/apps/CommandHistory');
+  var history, inputText;
+
+  beforeEach(function () {
+    history = new CommandHistory();
+    inputText = '';
+  });
+
+  it('moving back through history recounts commands in reverse order', function () {
```
Comment (davidsbailey) [inline apps/test/CommandHistoryTest.js]: this is totally optional, but part of the fun of this test framework is writing statements that kind of read like english. For example: `it('recounts commands in reverse order when moving back through history', ...`

File: apps/src/CommandHistory.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  /**
+   * The maximum number of entries to store in the command history, to
+   * prevent ever-increasing memory for this feature.
+   * @private {number}
+   */
+  this.maxEntries_ = 64;
+};
+
+/**
+ * Add the given command to the current command history.
```
Comment (bcjordan) [inline apps/src/CommandHistory.js]: Maybe also mention max entries expiration behavior

File: apps/src/CommandHistory.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+};
+
+/**
+ * Move back in time by one entry, returning the command at the new
+ * command index.
+ * @param {string} currentInput
+ * @returns {string}
+ */
+CommandHistory.prototype.goBack = function (currentInput) {
+  if (typeof this.commands_[this.currentIndex_]!== 'undefined') {
```
Comment (bcjordan) [inline apps/src/CommandHistory.js]: +space before `!==` (ah, noticed this was moved code)

File: apps/src/CommandHistory.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+};
+
+/**
+ * Move forward in time by one entry, returning the command at the new
+ * command index.
+ * @param {string} currentInput
+ * @returns {string}
+ */
+CommandHistory.prototype.goForward = function (currentInput) {
+  if (typeof this.commands_[this.currentIndex_]!== 'undefined') {
```
Comment (bcjordan) [inline apps/src/CommandHistory.js]: Noticed this was just moved too, could be good time to extract into a helper

File: shared/css/mixins.scss [hunk 5]
```diff
@@ -1,7 +1,7 @@
 // Useful mixins, especially for stuff that's browser-prefix heavy.
 // Could always use something like Prefixr for this instead...
 
-@mixin user-select($val) {
+@mixin user-selects($val) {
```
Comment (bcjordan) [inline shared/css/mixins.scss:4]: Looks good, regrettably I don't recall what the ultimate build issue ended up being with sprockets, we might have ruled this out. Think either form reads OK in any case.

File: apps/src/CommandHistory.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
+/** @file Provide browseable command history to a textbox with limited depth. */
```
Comment (bcjordan) [inline apps/src/CommandHistory.js]: Nit: deferring to Google

image

**Comment by bcjordan** [comment]
This is awesome, LGTM! Nothing but nits. Love the click-to-focus behavior, nice subtle enhancement.

File: shared/css/mixins.scss (hunk 5 above)
Comment (islemaster) [inline shared/css/mixins.scss:4]: Yeah, I'm not sure we ever figured out what the actual build issue was, just worked around it.  I'd like to give it another shot - being able to deduplicate this stuff is why we're using SCSS, right? (Never mind that we should probably be using [Prefixr](http://prefixr.com/) for this...)

**Comment by islemaster** [comment]
Merged in the interest of keeping things moving. @ty-po I'm still interested in your thoughts on this whenever you're available :+1:

**Pull request merged**